### ISSUE TITLE
Efficient Handling of Post Processing of Model Scores

### ISSUE THREAD

**Issue by Yuvraj-Dhepe** [Yuvraj-Dhepe/dcase2023-audio-retrieval#1]
During the model evaluation phase, we utilize the [postprocessing module](https://github.com/Yuvraj-Dhepe/dcase2023-audio-retrieval/tree/master/postprocessing). 
- [01_scores_wandb_new.py](https://github.com/Yuvraj-Dhepe/dcase2023-audio-retrieval/blob/master/postprocessing/01_scores_wandb_new.py): In this module we utilize the models to calculate cross-modal scores, across the text and audio
   - General idea is we compute score for an (audio,caption) for all the text captions in that particular split, likewise, we also calculate scores for a caption, taken in pair with all the audios
   - For ex. If train split has 7678 audios and each has 5 captions, then we will have about 23034 captions, which will be paired with 1 audio. Vice-versa, 1 caption will be paired with 7678 audios
   - These pairs will be used to evaluate audio2txt scores of R1, R5, R10 and mAP10 
- [02_xmodel_data_split.py](https://github.com/Yuvraj-Dhepe/dcase2023-audio-retrieval/blob/master/postprocessing/02_xmodel_data_split.py): In this module we split the audio2text (fid2items) and text2audio (tid2items) pairs into separate databases, for efficient handling in score calculations
- [03_retrieval_split_wandb_new.py](https://github.com/Yuvraj-Dhepe/dcase2023-audio-retrieval/blob/master/postprocessing/03_retrieval_split_wandb_new.py): We utilize this to calculate the metrics for txt2audio and audio2text using the fid2items.db and tid2items.db

The real issue arises, when we create 2 dbs for fid2items and tid2items in 
https://github.com/Yuvraj-Dhepe/dcase2023-audio-retrieval/blob/00602e761d8e81f7c7654520b44f8f8b6ed83d91/postprocessing/02_xmodel_data_split.py#L61-L87

- The reason for inefficiency is we need to keep the all_tid2items variable in memory, while whole processing is not completed. 
- Batching is not feasible, because we have to collect all fids for a single tid, and in batching it wouldn't work, cause the value for tid's will overwrite the old batches contents.
-  I tried a simple version, where I write tid_2_items where batch size is 1, so writing to tid_stream tid by tid, this gets very bad in terms of time consumption, like around 12 hrs just to process the tids (N I am talking about only 1x times the original data, I have to go to 5x times)
   - https://github.com/Yuvraj-Dhepe/dcase2023-audio-retrieval/blob/00602e761d8e81f7c7654520b44f8f8b6ed83d91/postprocessing/02_xmodel_data_split_batching.py#L70-L92 
- Even if I do it batching, like collecting some tids it still is quite slow 
   - https://github.com/Yuvraj-Dhepe/dcase2023-audio-retrieval/blob/3751491e4e9666e64af1cefee6cbaacf601f4792/postprocessing/02_xmodel_data_split_batching_2.py#L107-L131

Hence I would like to know how to handle this scenario in a better manner, so I can efficiently calculate the metrics, from the model scores. Currently, I only face the memory limitation to process the train set's model scores (**where train set is 3x or more than the original train set**) to get the evaluation metrics.

**Comment by Yuvraj-Dhepe** [comment]
[Validation Dataset for (2x original dataset) Model Scores](https://easyupload.io/ozlq5p). 
[Train Dataset for (2x original dataset) Model Scores](https://easyupload.io/ezc18s)

One can play with the dataset, and the scripts mentioned above to efficiently process them and get approximately the following scores using the retrieval files:

### Text to Audio Scores

| Split             | R1           | R5           | R10           | mAP           |
|-------------------|--------------|--------------|---------------|---------------|
| **eval**          | 0.132440191  | 0.371674641  | 0.500287081   | 0.232714665   |
| **val**           | 0.13645933   | 0.363444976  | 0.50277512    | 0.235242196   |
| **train**         | 0.13850333   | 0.350872623  | 0.468276709   | 0.229974988   |


### Audio to Text Scores

|      Split        | R1           | R5           | R10           | mAP           |
|-------------------|--------------|--------------|---------------|---------------|
| **eval**          | 0.025645933  | 0.119043062  | 0.191578947   | 0.098888661   |
| **val**           | 0.028133971  | 0.124210526  | 0.204210526   | 0.104337055   |
| **train**         | 0.063454024  | 0.266788226  | 0.382026569   | 0.159521473   |


There are just 1045 audio samples in validation set, so they were easily handled by above scripts, it's just the train set for which efficient memory management a need, when data becomes  **>=3x the original data.** Above metrics are for **2x the original data only for train set, rest valid and test data is same as original data**

**Comment by Yuvraj-Dhepe** [comment]
Solution is to not run evaluation only with eval and valid sets 😄

**Comment by Yuvraj-Dhepe** [comment]
Solution is to not run evaluation only with eval and valid sets 😄
